lib/FusionInventory/Agent/Tools/PartNumber.pm

```diff
@@ -0,0 +1,162 @@
+package FusionInventory::Agent::Tools::PartNumber;
+
+use strict;
+use warnings;
+
+use File::Glob;
+use English qw(-no_match_vars);
+
+use FusionInventory::Agent::Tools;
+use FusionInventory::Agent::Logger;
+
+use constant category       => "";
+use constant manufacturer   => "";
+
+# Lower priorities will be checked first
+use constant priority   => 50;
+
+my @subclasses;
+
+sub new {
+    my ($class, %params) = @_;
+
+    return unless defined($params{partnumber});
+
+    my $logger = $params{logger} || FusionInventory::Agent::Logger->new();
+
+    unless (@subclasses) {
+        my %priority = ();
+        my ($_classpath) = $INC{module2file(__PACKAGE__)} =~ /^(.*)\.pm$/;
+        foreach my $file (File::Glob::bsd_glob("$_classpath/*.pm")) {
+            $file =~ s{\\}{/}g if $OSNAME eq 'MSWin32';
+            my ($class) = $file =~ m/$_classpath\/(\S+)\.pm$/
+                or next;
+            my $module = "FusionInventory::Agent::Tools::PartNumber::" . $class;
+            $module->require();
+            if ($EVAL_ERROR) {
+                $logger->debug("Failed to load PartNumber::$class");
+                $logger->debug2("$module require error: $EVAL_ERROR");
+                next;
+            }
+            $priority{$module} = $module->priority();
+        }
+        @subclasses = sort { $priority{$a} <=> $priority{$b} } sort keys(%priority);
+    }
+
+    my $self = bless { _partnumber => $params{partnumber} }, $class;
+
+    foreach my $subclass (@subclasses) {
+        # Filter out by category and eventually by manufacturer
+        next if $params{category}     && $subclass->category     ne $params{category};
+        next if $params{manufacturer} && $subclass->manufacturer ne $params{manufacturer};
+        # Then match on regexp
+        my @matches = $params{partnumber} =~ $subclass->match_re
+            or next;
+        bless $self, $subclass;
+        $self->init(@matches);
+        last;
+    }
+
+    # Only validate Partnumber object if it has a manufacturer
+    return unless $self->manufacturer;
+
+    return $self;
+}
```

Comment:
Since you asked me to review I see many things i don't like :) :
* `new()` must always return an object regardless of parameters, imho.
* loading subclasses on every `new()` call is far from optimal. I would instantiate `FusionInventory::Agent::Tools::PartNumber` once with subclasses as "plugins" and store the resulting object in `$inventory`.

Reply:
* new() returns an object or nothing (`undef`) if no known manufacturer matches the provided string. The `undef` case permits to use a simple `if` on the returned value. This is perl-ish. Do you mean you would like to make a simple call like `if ($partnumber->known) {` ? where `known` is a method returning 0 by default ?
* why do you say subclasses are loaded on each `new` call ? In fact, there's a `@subclasses` class variable which make a cache of loaded subclasses. Perl modules are only parsed one time, did I miss something ? So subclass are only used to test the string after the first call.

Reply:
* This may be perl-ish, but not OOP-ish. I would make `$partnumber_factory->build(partnumber => $partnumber, manufacturer => $manufcaturer, category => $category)` which returns either an instance of a subclass or undef.
* To tell the truth I didn't notice @subclasses . This should work, but within a single caller module. You're also mixing OOP with non-OOP styles.